### ISSUE TITLE
API: STRtree interface consistently with shapely

### ISSUE THREAD

**Issue by jorisvandenbossche** [pygeos/pygeos#398]
We had some discussion in Shapely on changing the `STRtree` API in preparation for Shapely 2.0 / pygeos integration (https://github.com/Toblerity/Shapely/pull/1064, https://github.com/Toblerity/Shapely/pull/1112). And this eventually resulted in adding a `query_geoms` / `query_items` to return geometries and indices, respectively (instead of my original proposal to deprecate the current behaviour of returning geometries / adding a keyword to trigger this). The existing `query` was kept as alias of `query_geoms`.

And for the nearest query, we similarly now have `nearest_geom` and `nearest_item`.

I think it could be useful if we can already start adding methods to PyGEOS' STRtree that will be consistent with the future Shapely, so the eventual migration of pygeos to shapely 2.0 can be easier for the early adapters that are already using pygeos now. 
Alternatively, we can also keep it as is for now, and deal with those API changes at the moment when doing this pygeos->Shapely 2.0 migration (since that will involve some changes anyway, it might also be fine to wait until then with this). But we need to be prepared that there will be API changes at that point (or already start discussing that with other Shapely devs right now if we rather want to see changes in Shapely's API).

_If_ we already want to prepare for this, some questions:

- The `query` method clashes with the Shapely version. Do we add a `query_items` in pygeos? (and deprecate `query`?)
- The `query_bulk` concept doesn't yet exist in Shapely, so that's something we can keep as is (if the naming fits in the scheme of Shapely)
- The `nearest` and `nearest_all` methods have a different interface as Shapely's `nearest`/`nearest_item`, which return a single index and not an array of left+right indices. So in a sense, the pygeos method is more like a "nearest_bulk" ? (although I don't like that name)

**Comment by brendan-ward** [comment]
I don't know if it helps move this conversation forward, but this seems like it breaks down to 3 main aspects:
1. parameters or functions that expose different _behavior_ (or lower level functions that do different things, but exposed through a higher level API via keywords); e.g., return arbitrary equidistant vs all equidistant in the case of `nearest` vs `nearest_all`
2. parameters or functions that operate at a different level of granularity, e.g., scalar operations like `query` (aka `query_items` / `query_geometries`) vs vector `query_bulk` and `nearest` / `nearest_all` here
3. parameters or functions that control the type of thing that is returned: indices vs geometries, but based on the granularity and behavior from the above.

3 seems like it is entirely in the domain of the Python layer and what we call things is primarily a naming problem.

2 seems like a question of what things we want to implement in the C extensions vs Python.  I.e., we don't necessarily want to add another `nearest` function to the C extension that operates on a singular geometry equivalent to `query` here; we can handle this entirely in the Python tier and always pass in arrays to the C extension.  Likewise, we could eventually deprecate the `query` C method in favor of `query_bulk`, and then handle singular vs vector and geometry vs indices in the Python tier.

1 seems like the hardest problem since our underlying C extension methods provide different behaviors (`nearest` vs `nearest_all`) and has more decision points: closer mapping of pygeos function to GEOS function, or more custom behavior within pygeos.


If I follow correctly, it seems like most of what you are getting at here for future compatibility is naming?

**Comment by jorisvandenbossche** [comment]
> If I follow correctly, it seems like most of what you are getting at here for future compatibility is naming?

Indeed. The longer we wait with clearing this up, the more users already start using pygeos' methods, and the harder it will become to have them all change again when Shapely 2.0 comes out.

---

You are right that we don't necessarily have to implement all this in the C extension class. We already have a Python class wrapping the C extension class, and we can have more logic in that Python class to deal with the different cases (that's also what I am doing in https://github.com/Toblerity/Shapely/pull/1161, where I am basically making a version of such a Python class wrapping the C extension class for Shapely, and the ability to return a geometry instead of index is handled there in the Python level.

**Comment by caspervdw** [comment]
I am fine with adding the proposed names in PyGEOS and adding a deprecation warning at the .query method. 

Although the return value of query_bulk and nearest_all are similar, they do really different things. In fact “nearest” is already a bulk function. I think the naming and the API of these two are fine as is. Geometry-returning equivalent functions are not useful in these two cases.

Most/all of it can (and should) be done on the Python level.

**Comment by caspervdw** [comment]
@jorisvandenbossche Do you have the time to make a (draft) PR for this one?

**Comment by caspervdw** [comment]
As most things are now (almost) ready for 0.12, I think I’d like to postpone this, or solve it once we merge with shapely. Ok @jorisvandenbossche ?

**Comment by jorisvandenbossche** [comment]
Yes, let's maybe first resolve this in Shapely, and once that is done, we can still do a change in pygeos to add aliases / deprecation warnings.

**Comment by jorisvandenbossche** [comment]
Doing a first attempt in shapely: https://github.com/Toblerity/Shapely/pull/1251
